Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -2085,8 +2185,8 @@ protected class ClassWriter : IClassWriter {
             }
           });
         } else if (NeedsCustomReceiver(member) && !(member.EnclosingClass is TraitDecl)) {
-          // instance const in a newtype
-          Contract.Assert(typeArgs.Count == 0);
+          // instance const in a newtype or belongs to a datatype
+          Contract.Assert(typeArgs.Count == 0 || member.EnclosingClass is IndDatatypeDecl || member.EnclosingClass is CoDatatypeDecl);
```

Comment:
`member.EnclosingClass is IndDatatypeDecl || member.EnclosingClass is CoDatatypeDecl` can be just `member.EnclosingClass is DatatypeDecl`.